### PR TITLE
#47: Support for global tools

### PR DESCRIPTION
**Pull request by salimbouch**
addresses #47

### CODE REVIEW

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:40]: do not change the visibility of the field but instead of the `getMacOsHelper()` method.
Directly using this file in a sub-class would lead to NPEs as the getter lazily initializes the field.

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:436]: I would expect the `extract` logic to be available both for local and global tool commandlets.
While most global tools directly download an MSI, PKG, EXE, dpkg, etc. and install it without that we extract it in our Java code, there can be some edge-cases, where we will also need this for global tool commandlets in the future...

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:76]: this still looks like quite some redundancy between global and local tools that we could share in `ToolCommandlet`.

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:105]: There are two things to distinguish for global tools:
1. the installer binary used to install the tool globally on the system
2. the binary of the tool once it is installed on the system

According to the JavaDoc this method should refer to the 2nd one.
However, I can not see how we could implement this in a generic way.
Also we have to revisit the sense of this method after #111 - most probably this method will disappear (Sorry, when stories are implemented in parallel that have overlapping code changes it always gets kind of tricky and you will most probably also will have to deal with git conflicts. It therefore will make sense to work closer together with @aBega2000 so you both work together and not "against each other").

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:436]: BTW: the `isExtract()` method is there from the old design of devonfw-ide in bash and represents the parameter
`noUnpack` that was mainly used for global tool commandlets:
https://github.com/devonfw/ide/blob/af67497e06fb82e2e50a4d9ffa10e3f9dda44aa1/scripts/src/main/resources/scripts/functions#L1266

**Comment by salimbouch** [comment]
@hohwille do we want to have a file tool version in global tools? if so, should this file be next to the binary? if not, how do we know what version of the global tool we have and how can users update global tools?

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java]: Done, local are by default extractable and global arent but in need we can override isExtract and make it return true

Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java]: Question now is in case we need to extract a global tool where should we extract it? In the download cache or somewhere else?

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:104]: Im waiting for https://github.com/devonfw/IDEasy/pull/111 to be merged then I'll resolve the conflicts and change my code according to his changes.

Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:104]: I merged #111 now.

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java]: > Question now is in case we need to extract a global tool where should we extract it? In the download cache or somewhere else?

This shall be done in a temp folder as it will only be needed temporary to run the installer and after the installation (e.g. to `C:\Program Files`) it is not needed anymore. This is the same behavior as in devonfw-ide.

File: cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java:207]: never duplicate such code. Either move it from ToolCommandlet here or leave it in ToolCommandlet.

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:43]: If we need this for `doInstall`, you could simply use `this.context.getPath().findBinary(getBinaryName())`.
You have removed `getBinaryName()` from `ToolCommandlet`.
IMHO this must be reverted.

Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:43]: I removed it since it's the same as getName(), or is there something that I oversaw?

Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:43]: 1. Done.
2. Done. No need for `Files.isExecutable` now 
3. Done. That's exactly what I did but my commits weren't pushed.

File: cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java:206]: I made getInstalledVersion abstract in ToolCommandlet, implemented it in LocalToolCommandlet and in GlobalToolCommandlet it just gives null back.
Another solution would be to remove it entirely from ToolCommandlet and GlobalToolCommandlet and only implement it in LocalToolCommandlet then change GetVersionCommandlet to: 
`

    ToolCommandlet commandlet = this.tool.getValue();
    if (commandlet instanceof GlobalToolCommandlet) {
      throw new UnsupportedOperationException("Not implemented yet!");
    }
    VersionIdentifier installedVersion = ((LocalToolCommandlet)commandlet).getInstalledVersion();
    if (installedVersion == null) {
      throw new CliException("Tool " + commandlet.getName() + " is not installed!", 4);
    }
    this.context.info(installedVersion.toString());
  }`

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java]: 1- Should I replicate the overloading of installInRepo (i.e. installGlobally) or is it enough if I the whole logic is in doInstall?

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:371]: I did not realize before that I myself have duplicated this method. It is there twice:
* `public VersionIdentifier getInstalledVersion()` - fine
* `protected String getInstalledToolVersion()` - duplicated non-sense, it seems this is not used anywhere so it can simply be removed.

Sorry for getting a little off topic - not your actual task, but it would be great if we can already clean this up.
I also just found out that @aBega2000 forgot to add JavaDoc for his new method `getBinaryName()` that would also be nice to add.

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:36]: Never duplicate the JavaDoc for overridden methods (that is one of the reasons for `@Override` annotation).
The JavaDoc will always be inherited and all IDEs support this and also when you generate the JavaDoc as HTML this will happen. When you copy & paste the JavaDoc and it gets changed, the copy gets outdated and nobody will notice.

You may add additional information to the JavaDoc of inherited methods. But this is only to be done in very few cases.
Please note that JavaDoc describes the API of a method and not its implementation.
If you want to explain the implementation simply use code comments:
```

@Override
protected boolean isExtract() {
  // for global tools we usually download installers and do not want to extract them (e.g. installer.msi file shall not be extracted)
  return false;
}
```

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:49]: remove JavaDoc

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:97]: remove entirely. This class is abstract. If it can not implement this methods it should leave the task for the sub-classes to implement it.

File: cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java:209]: This method should be entirely removed. I already left a comment that this was a mistake on my end from the start.
Further do not implement methods "wrong". Either we do not support `getInstalledVersion()` for `GlobalToolCommdlet`s or we do. If we do then `GlobalToolCommdlet` can not easily implement this in a generic way. It would need to run the binary and ask it for its version like this:
```
$ docker -v
Docker version 24.0.5, build ced0996
```
Then the method has to be implemented in `DockerCommandlet` accordingly.

IMHO this makes sense as e.g. `VersionGetCommandlet` should also work for global tools.

BTW: your suggestion with `instanceof` is not making use of polymorphism. The beauty of OOP is that it simplifies such things. So if the behavior of your example would actually be desired, it would still be better to simply have the following in `GlobalToolCommandlet`:
```
@Override
protected String getInstalledToolVersion() {
  throw new UnsupportedOperationException("Not implemented yet!");
}
```
and in `LocalToolCommandlet` keep the rest of the implementation.
The `instanceof` approach makes `ToolCommandlet` to know `GlobalToolCommandlet` what otherwise is not required and more elegant.

Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java:209]: And finally what would make sense is something like this (most probably in `ToolCommandlet`) as it allows proper reuse:
```
  protected VersionIdentifier getInstalledVersion(Path toolPath) {
    
    if (!Files.isDirectory(toolPath)) {
      this.context.debug("Tool {} not installed in {}", getName(), toolPath);
      return null;
    }
    Path toolVersionFile = toolPath.resolve(IdeContext.FILE_SOFTWARE_VERSION);
    if (!Files.exists(toolVersionFile)) {
      Path legacyToolVersionFile = toolPath.resolve(IdeContext.FILE_LEGACY_SOFTWARE_VERSION);
      if (Files.exists(legacyToolVersionFile)) {
        toolVersionFile = legacyToolVersionFile;
      } else {
        this.context.warning("Tool {} is missing version file in {}", getName(), toolVersionFile);
        return null;
      }
    }
    try {
      String version = Files.readString(toolVersionFile).trim();
      return VersionIdentifier.of(version);
    } catch (IOException e) {
      throw new IllegalStateException("Failed to read file " + toolVersionFile, e);
    }
  }
```
Now we can easily implement this making reuse:
```
  public VersionIdentifier getInstalledVersion() {

    return getInstalledVersion(getToolPath());
  }
```

The benefit is that we can later also reuse `getInstalledVersion(Path)` to get the version of a tool from the software repository (e.g. to verify that all is sane).

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:281]: if you are moving methods within the same file, it looks like `isExtract()` or `getBinaryName()` has actually been added or removed but actually nothing has changed with these two methods at all.
I know that this quickly happens by accident and you do not notice and see this in your IDE.
However, always watch your diffs before you commit and also have a look into your diff in the PR to see that all is as expected and intended. It is not a big deal if this happens a few time. I just want to raise awareness to make life of reviewers easier in the future :)

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java]: A common pattern in OOP is also to have a template implementation in an abstract class like this:
```
public abstract class ChatBase {
  public void doSomething() {
    // common beginning part
    System.out.println("Hello, how are your?");
    // special part that differs
    doSomethingSpecial();
    // common ending part
    System.out.println("It was nice talking to you, goodbye.");
  }
  
  protected abstract void soSomethingSpecial();
}

public class ChatGoodMood extends ChatBase {
  @Override
  protected void soSomethingSpecial() {
    System.out.println("It is a beautiful day and I feel great.");
  }
}

public class ChatBadMood extends ChatBase {
  @Override
  protected void soSomethingSpecial() {
    System.out.println("The wars are not ending and the climate will collapse. I feel very sad.");
  }
}
```

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java]: Method is gone, resolved.

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:371]: Done

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:36]: Done. Thanks for clarifying

Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:36]: I removed `Override it to add custom post installation logic.` from the javaDoc of the method postInstall.

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:49]: Done

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java:97]: Done

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:281]: I will make sure to avoid this in my next PRs :)

File: cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java:209]: Done. Thanks for the clarification.

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:184]: Did you remove `Override it to add custom post intallation logic.` on purpose or by accident?

**Review by hohwille — APPROVED** [review]
@salimbouch Thanks for your rework. Now everything is perfect. As I had quite a list of review comments and added improvements it is especially great to see that you solved it all in one shot 👍 
Ready for merge.

The last comment is just for my curiosity and shall not block the merge...

File: cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/GlobalToolCommandlet.java]: IMHO we can do such nice-to-have improvements after the merge.
It is better to have this already on main branch so it gets available for other stories already worked on like #75.

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/tool/ToolCommandlet.java:184]: I just saw your explanation in the other comment...

**Pull request merged**
